src/plugins/kibana_react/public/code_editor/code_editor.tsx

```diff
@@ -102,116 +110,244 @@ export interface Props {
     }
+  }, []);
+
+  const startEditing = useCallback(() => {
+    setIsHintActive(false);
+    _editor.current?.focus();
+  }, []);
+
+  const stopEditing = useCallback(() => {
+    setIsHintActive(true);
+  }, []);
+
+  const onKeyDownHint = useCallback(
+    (ev) => {
+      if (ev.keyCode === keyCodes.ENTER) {
+        ev.preventDefault();
+        startEditing();
+      }
+    },
+    [startEditing]
+  );
+
+  const onKeydownMonaco = useCallback(
+    (ev) => {
+      // Make sure the textarea is not directly accesible with TAB
+      const textbox = ev.target;
+      textbox.tabIndex = -1;
+
+      if (ev.keyCode === keyCodes.ESCAPE) {
+        // If the autocompletion context menu is open then we want to let ESCAPE close it but
+        // **not** exit out of editing mode.
+        if (!isSuggestionMenuOpen.current) {
+          ev.preventDefault();
+          ev.stopPropagation();
+          stopEditing();
+          editorHint.current?.focus();
+        }
+      }
+    },
+    [stopEditing]
+  );
 
-    if (this.props.editorWillMount) {
-      this.props.editorWillMount();
-    }
+  const renderPrompt = useCallback(() => {
+    return (
+      <div
+        className={promptClasses}
+        id={htmlIdGenerator('codeEditor')()}
+        ref={editorHint}
+        tabIndex={0}
+        role="button"
+        onClick={startEditing}
+        onKeyDown={onKeyDownHint}
+        data-test-subj="codeEditorHint"
+      >
+        <EuiText>
+          <p>
+            <FormattedMessage
+              id="kibana-react.kibanaCodeEditor.startEditing"
+              defaultMessage="Press Enter to start editing."
+            />
+          </p>
+        </EuiText>
+
+        <EuiText>
+          <p>
+            <FormattedMessage
+              id="kibana-react.kibanaCodeEditor.stopEditing"
+              defaultMessage="When you're done, press Escape to stop editing."
+            />
+          </p>
+        </EuiText>
+      </div>
+    );
+  }, [onKeyDownHint, promptClasses, startEditing]);
 
-    monaco.languages.onLanguage(this.props.languageId, () => {
-      if (this.props.suggestionProvider) {
-        monaco.languages.registerCompletionItemProvider(
-          this.props.languageId,
-          this.props.suggestionProvider
-        );
+  const _editorWillMount = useCallback(
+    (__monaco: unknown) => {
+      if (__monaco !== monaco) {
+        throw new Error('react-monaco-editor is using a different version of monaco');
       }
 
-      if (this.props.signatureProvider) {
-        monaco.languages.registerSignatureHelpProvider(
-          this.props.languageId,
-          this.props.signatureProvider
-        );
+      if (overrideEditorWillMount) {
+        overrideEditorWillMount();
+        return;
       }
 
-      if (this.props.hoverProvider) {
-        monaco.languages.registerHoverProvider(this.props.languageId, this.props.hoverProvider);
+      if (editorWillMount) {
+        editorWillMount();
       }
```

Comment:
:art: codestyle nitpick

```suggestion
      editorWillMount?.();
```